Fix skiplink target so it doesn't take up space and cause overflow

yvonnetangsu opened the pull request:
# READY FOR REVIEW

# Summary
- The skiplink target was using the `.su-sr-only-text` class which uses the Bourbon `@hide-text` mixin which takes up space and causes overflow. This PR creates new class `.su-skiplinks__target` which uses the `@hide-visually` mixin instead which has the style code we need. The skip link target now has proper styles (same as current best practice).
- I didn't just use the `.su-sr-only-element` class on the skip link target directly because .su-sr-only-element unhides on focus.
- Add documentation for `.su-sr-only-text` - it's only good for cases where e.g., hide the SR text in the site-search component magnifying glass button/icon.

# Needed By (Date)
- ASAP

# Urgency
- Pretty urgent

# Steps to Test
1. Pull this branch and run `npm run build`.
2. Look at the skiplink target in the skiplinks component. It should now now take up any spaces.


# Affected Projects or Products
- Decanter

# Associated Issues and/or People
- #414

JBCSU commented on core/src/scss/components/skiplinks/_skiplinks.scss:
```diff
@@ -19,7 +19,7 @@
 // </header>
 //  <main>
-//   <h2 id="main-content" class="su-sr-only-text" tabindex="-1">Main Content</h2>
+//   <h2 id="main-content" class="su-skiplinks__target" tabindex="-1">Main Content</h2>
```
Why do we need a new class? Why can't we just fix `.su-sr-only-text`?

If we do need a new class, I'd prefer to not use a name that implies you can only use the class on one item, e.g. skiplinks. I would expect the class to be useful for anything that needs to be hidden, e.g. the supplemental text in **Read more** links. Is there some reason the skip link target need to be treated specially?

yvonnetangsu replied:
.su-sr-only-text uses the Bourbon mixin @hide-text which is good for some use cases, eg, for the SR only text in site search magnifying glass icon so I want to preserve that.

We currently have 2 existing SR helper classes (the above mentioned .su-sr-only-text, and .su-sr-only-element) and neither of them function the way we want the skiplink target to function. One way we could make this solution more general would be to rethink .su-sr-only-element. Currently, it unhides when it's on-focus. I don't believe any components in Decanter uses it, but this class might be used by some other sites that use Decanter? This might be something that needs a discussion.

JBCSU replied:
Are you saying your implementation, `@hide-visually`, wouldn't work for the search icon?

yvonnetangsu replied:
Yeah, I believe that's what I tried initially and it didn't work.

JBCSU replied:
Ok, so we have a few issues here:

Apparently the point of `.su-sr-only-text` is to hide just the textual content of an element, but render everything else, e.g. background, border, padding, .... `.su-sr-only-element` is to hide the entire element, including background, border, padding, .... Applying `.su-sr-only-element` to the search icon hides the icon as well as the helper text, because the icon is a background image for the helper text.

In testing this, I found the first issue, which is that the search icon's helper text is never read aloud by voice over - the `aria-label="Search"` is read, regardless of what text is provided for the `<button>`. So that helper text is completely ignored, at least by voice over - we can remove it entirely and therefore we can remove `.su-sr-only-text` from the button.

If the target of the skip link is, as we recommend, an `<h2>` whose sole purpose is to announce to screen-readers "Main content", then we want to hide the entire element. So `.su-sr-only-element` seems appropriate. So now we need to decide what should happen when focus lands on the sr-only-element.

I propose that `.su-sr-only-element` does not reveal on focus, and that we add a variant, `.su-sr-only-element--reveal-on-focus`, that does.

That said, I think the target of the skip link, which is the aforementioned `<h2>`, actually should reveal on focus. When I'm testing keyboard nav and I hit the skip link, I'm left wondering where my focus went. I have to keep tabbing to figure out where I'm at. I think if the `<h2>Main Content</h2>` revealed itself it would be as useful to keyboard navigators as it is to screen readers. Jiatyan agrees that that would be helpful.

yvonnetangsu replied:
I'm curious about the aria-label/button text issue also. If Level Access can confirm that it is ok to remove the button text entirely then we can certainly just do that.

I'm open to the idea of skip link targets reveals itself on focus. I also agree that our default `.su-sr-only-element` should not reveal on focus and I like the idea of a variant that does reveal itself on focus.

Other Decanter people - any thoughts? 😃

sherakama replied:
Might as well use the bourbon mixins for this. As the target is its own element I would suggest changing the name to `su-skiplinks-target` instead of the BEM child as it really isn't nested. If you want to reveal the element on focus you can use the unhide paramater as suggested in the bourbon documentation.

```
/// @example scss
///   .su-skiplinks-target {
///     @include hide-visually;
///
///     &:active,
///     &:focus {
///       @include hide-visually("unhide");
///     }
///   }
```

sherakama replied:
As this is pretty much all new it doesn't break compatibility and we can safely merge and go on.

sherakama commented:
Original issue here: https://github.com/SU-SWS/decanter/issues/414

yvonnetangsu commented:
> As the target is its own element I would suggest changing the name to `su-skiplinks-target` instead of the BEM child as it really isn't nested.
> 
> Other than that this looks good to me.

Thanks for the review, @sherakama ! I had a discussion with JB last week and she's not keen on the idea of adding a new class name so instead she'd rather we tweak the default `.su-sr-only-element` class and remove the unhide on focus behavior, and adding a variant to `.su-sr-only-element` class that does unhide on focus. @JBCSU - comments? Would that work, @sherakama?

yvonnetangsu commented:
@sherakama , @JBCSU , I removed the su-skiplinks__target class, used the su-sr-only-element class instead and changed the su-sr-only-element helper class so it doesn't unhide on focus. One thing I didn't do is putting unhide on focus in the skiplink target - I remember @JBCSU , Scott and I had a discussion and he preferred that skiplink targets do not unhide on focus, since that's the conventional thing out there for skiplinks and he didn't want to confuse keyboard users?
